src/components/account/MarketTopUp.tsx

```diff
@@ -0,0 +1,118 @@
+import { useState } from "react";
+import { useCtx } from "../../GlobalCtx";
+import { Transaction, TransactionState, type TransactionStatus } from "../../lib/transactionStatus";
+
+interface MarketTopUpProps {
+  selectedAddress: string;
+  walletBalance: bigint;
+  onSuccess: () => void;
+}
+
+export function MarketTopUpPanel({ selectedAddress, walletBalance, onSuccess }: MarketTopUpProps) {
+  const { collatorWsApi: api } = useCtx();
+  const [topUpAmount, setTopUpAmount] = useState("");
+  const [topUpStatus, setTopUpStatus] = useState<TransactionStatus>(Transaction.idle);
+
+  const handleTopUp = async () => {
+    if (!api || !selectedAddress || !topUpAmount) return;
+
+    try {
+      setTopUpStatus(Transaction.loading);
+
+      const unsub = await api.tx.market
+        .addBalance(topUpAmount)
+        .signAndSend(selectedAddress, ({ status, dispatchError, txHash }) => {
+          if (!status.isInBlock && !status.isFinalized) {
+            return;
+          }
+          try {
+            if (!dispatchError) {
+              const txHashHex = txHash.toHex();
+              setTopUpStatus(Transaction.success(txHashHex));
+              onSuccess();
+              return;
+            }
+            if (!dispatchError.isModule) {
+              setTopUpStatus(Transaction.error(dispatchError.toString()));
+              return;
+            }
+            const decoded = api.registry.findMetaError(dispatchError.asModule);
+            const { docs, name, section } = decoded;
+            const userMessage = docs.join(" ") || `${section}.${name}`;
+
+            setTopUpStatus(Transaction.error(userMessage));
+          } finally {
+            unsub();
+          }
+        });
+    } catch (err) {
+      console.error("Balance top up failed", err);
+      if (err instanceof Error) {
+        const isDuplicate =
+          err.message.includes("1013") || err.message.includes("Transaction Already Imported");
+
+        const userMessage = isDuplicate
+          ? "Your transaction is already being processed."
+          : "Transaction failed.";
+
+        setTopUpStatus(Transaction.error(userMessage));
+      } else {
+        setTopUpStatus(Transaction.error("Failed to decode incoming error, see logs for details."));
```

Comment:
We don't need to change this now, but IMO this kind of logic should be kept outside the UI